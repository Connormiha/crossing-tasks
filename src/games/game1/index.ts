import characters from './characters';

import {RIVERSIDE_LEFT, RIVERSIDE_RIGHT, BOAT} from 'flux/types';

import {
    Game
} from 'games/helpers';

export default new Game(
    characters,
    {
        [BOAT]: [],
        [RIVERSIDE_LEFT]: characters.map(({id}) => id),
        [RIVERSIDE_RIGHT]: []
    },
    'Crossing 1',
    'It is necessary to get everyone across the river.',
    {
        beforeLanding: [
            {
                description: 'The boat can accommodate only two persons',
                check(collocation, _, moveTo): boolean {
                    return moveTo !== BOAT || collocation.boat.length < 2;
                }
            }
        ],
        beforeDeparture: [
            {
                description: 'Driving the boat can only farmer',
                check(collocation): boolean {
                    return collocation.boat.includes('farmer');
                }
            },
            {
                description: 'Do not leave a wolf and a sheep farmer without one',
                check(collocation): boolean {
                    for (const item of [RIVERSIDE_LEFT, RIVERSIDE_RIGHT]) {
                        const side = collocation[item];

                        if (!side.includes('farmer') && side.includes('wolf') && side.includes('sheep')) {
                            return false;
                        }
                    }

                    return true;
                }
            },
            {
                description: 'Do not leave the sheep and cabbage alone without the farmer',
                check(collocation): boolean {
                    for (const item of [RIVERSIDE_LEFT, RIVERSIDE_RIGHT]) {
                        const side = collocation[item];

                        if (!side.includes('farmer') && side.includes('cabbage') && side.includes('sheep')) {
                            return false;
                        }
                    }

                    return true;
                }
            }
        ]
    },
);
