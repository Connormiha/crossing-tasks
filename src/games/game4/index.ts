import characters from './characters';

import {
    RIVERSIDE_LEFT, RIVERSIDE_RIGHT, BOAT,
    Game, landingValidator, depetureValidator
} from 'games/helpers';

const game: Game = {
    title: 'Crossing 4',
    description: '@todo',
    rules: {
        beforeLanding: [
            {
                description: 'The boat can accommodate only two persons',
                check(collocation: any, characterId: string, moveTo: string): boolean {
                    return moveTo !== BOAT || collocation.boat.length < 2;
                }
            }
        ],
        beforeDeparture: [
            {
                description: 'The boat is empty',
                check(collocation: any): boolean {
                    return collocation.boat.length !== 0;
                }
            },
            {
                description: 'Driving the boat can only adult',
                check(collocation: any): boolean {
                    for (let character of collocation.boat) {
                        if (characters[character].adult) {
                            return true;
                        }
                    }

                    return false;
                }
            },
            {
                description: 'The criminal shoudn\'t stay with citizens in boat',
                check(collocation: any): boolean {
                    if (collocation.boat.length === 2 &&
                        !collocation.boat.includes('policeman') &&
                        collocation.boat.includes('criminal')
                    ) {
                        return false;
                    }

                    return true;
                }
            },
            {
                description: 'The criminal shoudn\'t stay with citizens in riverside',
                check(collocation: any): boolean {
                    for (let item of [RIVERSIDE_LEFT, RIVERSIDE_RIGHT]) {
                        let side = collocation[item];

                        if (side.includes('criminal') &&
                            !side.includes('policeman') &&
                            side.length > 1
                        ) {
                            return false;
                        }
                    }

                    return true;
                }
            },
            {
                description: 'Boys shoudn\'t stay with mother in riverside or boat alone without father',
                check(collocation: any): boolean {
                    for (let item of [RIVERSIDE_LEFT, RIVERSIDE_RIGHT, BOAT]) {
                        let side = collocation[item];

                        if (item !== collocation.boatPosition && collocation.boatPosition !== BOAT) {
                            side = side.concat(collocation[BOAT]);
                        }

                        if (side.includes('woman_black') &&
                            !side.includes('men_black') &&
                            (side.includes('boy_yellow') || side.includes('boy_red'))
                        ) {
                            return false;
                        }
                    }

                    return true;
                }
            },
            {
                description: 'Girls shoudn\'t stay with father in riverside or boat alone without mother',
                check(collocation: any): boolean {
                    for (let item of [RIVERSIDE_LEFT, RIVERSIDE_RIGHT, BOAT]) {
                        let side = collocation[item];

                        if (item !== collocation.boatPosition && collocation.boatPosition !== BOAT) {
                            side = side.concat(collocation[BOAT]);
                        }

                        if (side.includes('men_black') &&
                            !side.includes('woman_black') &&
                            (side.includes('girl_yellow') || side.includes('girl_red'))
                        ) {
                            return false;
                        }
                    }

                    return true;
                }
            }
        ]
    },
    landingValidator,
    depetureValidator,
    characters,
    collocation: {
        [BOAT]: [],
        [RIVERSIDE_LEFT]: Object.keys(characters),
        [RIVERSIDE_RIGHT]: []
    }
};

export default game;
