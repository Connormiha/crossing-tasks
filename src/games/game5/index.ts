import characters from './characters';

import {RIVERSIDE_LEFT, RIVERSIDE_RIGHT, BOAT} from 'flux/types';

import {
    Game,
} from 'games/helpers';

const game = new Game(
    characters,
    {
        [BOAT]: [],
        [RIVERSIDE_LEFT]: characters.map(({id}) => id),
        [RIVERSIDE_RIGHT]: []
    },
    'Crossing 5',
    'It is necessary to get everyone across the river.',
    {
        beforeLanding: [
            {
                description: 'The boat can accommodate only two persons',
                check(collocation, _, moveTo): boolean {
                    return moveTo !== BOAT || collocation.boat.length < 2;
                }
            }
        ],
        beforeDeparture: [
            {
                description: 'The boat is empty',
                check(collocation): boolean {
                    return collocation.boat.length !== 0;
                }
            },
            {
                description: 'Only big monkey or human can drive the boat',
                check(collocation): boolean {
                    for (const character of collocation[BOAT]) {
                        if (character === 'gorilla' || characters[character].type === 'priest') {
                            return true;
                        }
                    }

                    return false;
                }
            },
            {
                description: 'Monkeys do not have to get numerical superiority in any riverside',
                check(collocation): boolean {
                    for (const item of [RIVERSIDE_LEFT, RIVERSIDE_RIGHT]) {
                        let side: string[] = collocation[item],
                            priestsCount = 0,
                            monkeyCount = 0;

                        if (item !== collocation.boatPosition) {
                            side = side.concat(collocation[BOAT]);
                        }

                        for (const character of side) {
                            if (characters[character].type === 'priest') {
                                priestsCount++;
                            } else {
                                monkeyCount++;
                            }
                        }

                        if (priestsCount && priestsCount < monkeyCount) {
                            return false;
                        }
                    }

                    return true;
                }
            }
        ]
    },
);

export default game;
