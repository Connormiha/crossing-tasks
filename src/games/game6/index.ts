import characters from './characters';

import {RIVERSIDE_LEFT, RIVERSIDE_RIGHT, BOAT} from 'flux/types';

import {
    Game, landingValidator, depetureValidator
} from 'games/helpers';

const game: Game = {
    title: 'Crossing 3',
    description: 'No woman with other men without her hasband (or other woman)',
    rules: {
        beforeLanding: [
            {
                description: 'The boat can accommodate only two persons',
                check(collocation: any, characterId: string, moveTo: string): boolean {
                    return moveTo !== BOAT || collocation.boat.length < 2;
                }
            }
        ],
        beforeDeparture: [
            {
                description: 'The boat is empty',
                check(collocation: any): boolean {
                    return collocation.boat.length !== 0;
                }
            },
            {
                description: 'A woman can not be with a man without a husband in boat',
                check(collocation: any): boolean {
                    let side = collocation[BOAT],
                        mens: string[] = [],
                        women: string[] = [];

                    for (let character of side) {
                        if (characters[character].sex === 'female') {
                            women.push(characters[character].family);
                        } else {
                            mens.push(characters[character].family);
                        }
                    }

                    if (mens.length) {
                        for (let woman of women) {
                            if (!mens.includes(woman)) {
                                return false;
                            }
                        }
                    }

                    return true;
                }
            },

            {
                description: 'A woman can not be with a men without a husband or other woman in riverside',
                check(collocation: any): boolean {
                    for (let item of [RIVERSIDE_LEFT, RIVERSIDE_RIGHT]) {
                        let side: string[] = collocation[item],
                            mens: string[] = [],
                            women: string[] = [];

                        if (item !== collocation.boatPosition) {
                            side = side.concat(collocation[BOAT]);
                        }

                        for (let character of side) {
                            if (characters[character].sex === 'female') {
                                women.push(characters[character].family);
                            } else {
                                mens.push(characters[character].family);
                            }
                        }

                        if (mens.length &&
                            women.length === 1 &&
                            !mens.includes(women[0])
                        ) {
                            return false;
                        }
                    }

                    return true;
                }
            }
        ]
    },
    landingValidator,
    depetureValidator,
    characters,
    collocation: {
        [BOAT]: [],
        [RIVERSIDE_LEFT]: Object.keys(characters),
        [RIVERSIDE_RIGHT]: []
    }
};

export default game;
