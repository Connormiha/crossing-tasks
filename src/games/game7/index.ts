import characters from './characters';

import {RIVERSIDE_LEFT, RIVERSIDE_RIGHT, BOAT} from 'flux/types';

import {
    Game, landingValidator, depetureValidator
} from 'games/helpers';

const game: Game = {
    title: 'Crossing 7',
    description: 'It is necessary to get everyone across the river.',
    rules: {
        beforeLanding: [
            {
                description: 'The boat can accommodate only 3 persons',
                check(collocation: any, characterId: string, moveTo: string): boolean {
                    return moveTo !== BOAT || collocation.boat.length < 3;
                }
            }
        ],
        beforeDeparture: [
            {
                description: 'Driving the boat can only farmer',
                check(collocation: any): boolean {
                    return collocation.boat.includes('farmer');
                }
            },
            {
                description: 'Do not leave a wolf and a sheep alone without the farmer',
                check(collocation: any): boolean {
                    for (const item of [RIVERSIDE_LEFT, RIVERSIDE_RIGHT]) {
                        const side = collocation[item];

                        if (!side.includes('farmer') && (side.includes('wolf1') || side.includes('wolf2')) && side.includes('sheep')) {
                            return false;
                        }
                    }

                    return true;
                }
            },
            {
                description: 'Do not leave a wolf and a dog alone without the farmer',
                check(collocation: any): boolean {
                    for (const item of [RIVERSIDE_LEFT, RIVERSIDE_RIGHT]) {
                        const side = collocation[item];

                        if (!side.includes('farmer') && (side.includes('wolf1') || side.includes('wolf2')) && side.includes('dog')) {
                            return false;
                        }
                    }

                    return true;
                }
            },
            {
                description: 'Do not leave a dog and a sheep alone without the farmer',
                check(collocation: any): boolean {
                    for (const item of [RIVERSIDE_LEFT, RIVERSIDE_RIGHT]) {
                        const side = collocation[item];

                        if (!side.includes('farmer') && side.includes('dog') && side.includes('sheep')) {
                            return false;
                        }
                    }

                    return true;
                }
            },
            {
                description: 'Do not leave the sheep and cabbage alone without the farmer',
                check(collocation: any): boolean {
                    for (const item of [RIVERSIDE_LEFT, RIVERSIDE_RIGHT]) {
                        const side = collocation[item];

                        if (!side.includes('farmer') && side.includes('cabbage') && side.includes('sheep')) {
                            return false;
                        }
                    }

                    return true;
                }
            }
        ]
    },
    landingValidator,
    depetureValidator,
    characters,
    collocation: {
        [BOAT]: [],
        [RIVERSIDE_LEFT]: Object.keys(characters),
        [RIVERSIDE_RIGHT]: []
    }
};

export default game;
