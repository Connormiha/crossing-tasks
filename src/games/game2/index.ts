import characters from './characters';

import { RIVERSIDE_LEFT, RIVERSIDE_RIGHT, BOAT } from 'flux/types';

import { Game } from 'games/helpers';

export default new Game(
  characters,
  {
    [BOAT]: [],
    [RIVERSIDE_LEFT]: characters.map(({ id }) => id),
    [RIVERSIDE_RIGHT]: [],
  },
  'Crossing 2',
  'It is necessary to get everyone across the river.',
  {
    beforeLanding: [
      {
        description: 'The boat can accommodate only two persons',
        check(collocation): boolean {
          return collocation.boat.length < 2;
        },
      },
    ],
    beforeDeparture: [
      {
        description: 'The boat is empty',
        check(collocation): boolean {
          return collocation.boat.length !== 0;
        },
      },
      {
        description: 'Savages do not have to get numerical superiority in any riverside',
        check(collocation): boolean {
          for (const item of [RIVERSIDE_LEFT, RIVERSIDE_RIGHT]) {
            let side = collocation[item],
              priestsCount = 0,
              cavemansCount = 0;

            if (item !== collocation.boatPosition) {
              side = side.concat(collocation[BOAT]);
            }

            for (const character of side) {
              if (characters[character].type === 'priest') {
                priestsCount++;
              } else {
                cavemansCount++;
              }
            }

            if (priestsCount && priestsCount < cavemansCount) {
              return false;
            }
          }

          return true;
        },
      },
    ],
  },
);
