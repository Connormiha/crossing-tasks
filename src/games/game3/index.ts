import characters from './characters';

import { RIVERSIDE_LEFT, RIVERSIDE_RIGHT, BOAT } from 'flux/types';
import { ICharacterIdGame3, ICharacterFamily } from './types';

import { Game } from 'games/helpers';

const isWomanWithOtherMenWithuotHasband = (side: ICharacterIdGame3[]): boolean => {
  const mens: ICharacterFamily[] = [];
  const women: ICharacterFamily[] = [];

  for (const character of side) {
    const item = characters.find(({id}) => id === character)!;

    if (item.sex === 'female') {
      women.push(item.family);
    } else {
      mens.push(item.family);
    }
  }

  if (mens.length) {
    for (const woman of women) {
      if (!mens.includes(woman)) {
        return false;
      }
    }
  }

  return true;
};

export default new Game(
  characters,
  {
    [BOAT]: [],
    [RIVERSIDE_LEFT]: characters.map(({ id }) => id),
    [RIVERSIDE_RIGHT]: [],
  },
  'Crossing 3',
  'No woman with other men without her hasband',
  {
    beforeLanding: [
      {
        description: 'The boat can accommodate only two persons',
        check(collocation, _, moveTo): boolean {
          return moveTo !== BOAT || collocation.boat.length < 2;
        },
      },
    ],
    beforeDeparture: [
      {
        description: 'The boat is empty',
        check(collocation): boolean {
          return collocation.boat.length !== 0;
        },
      },
      {
        description: 'A woman can not be with a man without a husband in boat',
        check(collocation): boolean {
          return isWomanWithOtherMenWithuotHasband(collocation[BOAT]);
        },
      },

      {
        description: 'A woman can not be with a man without a husband in riverside',
        check(collocation): boolean {
          for (const item of [RIVERSIDE_LEFT, RIVERSIDE_RIGHT]) {
            const side: ICharacterIdGame3[] = collocation[item];

            if (item !== collocation.boatPosition) {
              side.push(...collocation[BOAT]);
            }

            if (!isWomanWithOtherMenWithuotHasband(side)) {
              return false;
            }
          }

          return true;
        },
      },
    ],
  },
);
