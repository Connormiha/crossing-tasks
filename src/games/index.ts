import game1 from 'games/game1';
import game2 from 'games/game2';
import game3 from 'games/game3';
import game4 from 'games/game4';
import game5 from 'games/game5';
import game6 from 'games/game6';
import game7 from 'games/game7';

import {
    Game, ValidatorResult, PositionCharacter
} from 'games/helpers';

export {
    Game, ValidatorResult, PositionCharacter
};

const games = {
    game1, game2, game3, game4, game5, game6, game7
};

export default games;
