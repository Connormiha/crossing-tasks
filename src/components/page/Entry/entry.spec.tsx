import * as React from 'react';
import * as renderer from 'react-test-renderer';
import {BrowserRouter as Router} from 'react-router-dom';
import PageEntryPure from 'components/page/entry/index.pure';
import games from 'games';

/* tslint:disable:jsx-wrap-multiline */
describe('<PageEntryPure />', () => {
    it('should render', () => {
        const tree = renderer.create(
            <Router>
                <PageEntryPure
                    game={{
                        list: Object.keys(games),
                        finished: false,
                        currentGame: 'game1'
                    }}
                />
            </Router>
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });
});
