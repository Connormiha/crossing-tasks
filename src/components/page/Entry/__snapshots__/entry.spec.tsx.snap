// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PageEntryPure /> should render 1`] = `
<div
  className="page-entry"
>
  <div
    className="page-entry__content"
  >
    <ul
      className="list-games"
    >
      <li
        className="list-games__item"
      >
        <a
          href="/play/game1/"
          onClick={[Function]}
        >
          game1
        </a>
      </li>
      <li
        className="list-games__item"
      >
        <a
          href="/play/game2/"
          onClick={[Function]}
        >
          game2
        </a>
      </li>
      <li
        className="list-games__item"
      >
        <a
          href="/play/game3/"
          onClick={[Function]}
        >
          game3
        </a>
      </li>
      <li
        className="list-games__item"
      >
        <a
          href="/play/game4/"
          onClick={[Function]}
        >
          game4
        </a>
      </li>
      <li
        className="list-games__item"
      >
        <a
          href="/play/game5/"
          onClick={[Function]}
        >
          game5
        </a>
      </li>
      <li
        className="list-games__item"
      >
        <a
          href="/play/game6/"
          onClick={[Function]}
        >
          game6
        </a>
      </li>
      <li
        className="list-games__item"
      >
        <a
          href="/play/game7/"
          onClick={[Function]}
        >
          game7
        </a>
      </li>
    </ul>
  </div>
</div>
`;
